Render validation

DEVetter opened the pull request:
Implemented validation rules render-20201:render-23043 (i.e. missing the early rules render-10101:render-20103, also missing the recommendations render-21006/21007).

On the way, did some refactoring to the render-package. Some key refactorings:
 - dealing with incorrect attributes when reading: Added non-object-oriented helper-method `static void addToInvalidXMLUserObject` to XMLTools to register invalid attribute-values for later validation; and used it in the `readAttribute`-method
 - polygon's getChildAt can now handle the usual calls for notes/annotation (and is consistent with Polygon.getChildCount) -- this may need be translated to RenderCurve too, but RenderCurve did not cause problems here
 - Added UserObject-tracking in RenderParser, but for now just for RenderInformationBase (see respective TODO) -- this may need be added for the other Classes where subobjects are to be expected (cf. LayoutL3Parser)
 - Added `boolean isListOf___Empty()`-methods to complement the `boolean isSetListOf___()`-methods: `isListOf___Empty` will return `true` iff the list is not null, but empty (which is usually forbidden by the validation-rules). Thereby, the API as stood before does not change (`isSetListOf___` still behave the same).

The sbml-test-files were adapted where they (incorrectly?) made use of <g ... /> instead of <render:g ... />.

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/TransformationConstraints.java:
```diff
@@ -0,0 +1,87 @@
+        @Override // Any string
+        public boolean check(ValidationContext ctx, Transformation t) { return true; }
+      };
+      break;
```
Possibly forgotten `default: break;` here?

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/LocalStyleConstraints.java:
```diff
@@ -0,0 +1,86 @@
+          return true;
+        }
+      };
+      break;
```
Do we need a `default: break;` here?

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/PolygonConstraints.java:
```diff
@@ -0,0 +1,94 @@
+              LayoutConstants.listOfCurveSegments).check(ctx, poly);
+        }
+      };
+      break;
```
`default: break;`?

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/LinearGradientConstraints.java:
```diff
@@ -0,0 +1,131 @@
+            || gradient.getZ2().isSetRelativeValue();
+        }
+      };
+      break;
```
`default: break;`?

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/LocalRenderInformationConstraints.java:
```diff
@@ -0,0 +1,125 @@
+          return true;
+        }
+      };
+      break;
```
`default: break;`?

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/RenderInformationBaseConstraints.java:
```diff
@@ -0,0 +1,230 @@
+          return true;
+        }
+      };
+      break;
```
`default: break;`?

draeger commented on extensions/render/src/org/sbml/jsbml/validator/offline/constraints/StyleConstraints.java:
```diff
@@ -61,11 +70,34 @@ public void addErrorCodesForAttribute(Set<Integer> set, int level,
+        @Override // any string
+        public boolean check(ValidationContext ctx, Style t) { return true; }
+      };    
+      break;
```
`default: break;`?

draeger commented:
Don't we need `default: break;` statements for all `switch` statements?

DEVetter commented:
> Don't we need `default: break;` statements for all `switch` statements?
We did not, but now we do. I added them and moved the initialisation `func=null` to the default statement.

niko-rodrigue commented on extensions/render/src/org/sbml/jsbml/ext/render/XMLTools.java:
```diff
@@ -224,5 +246,32 @@ public static String encodeArrayShortToString(Short[] array) {
+   */
+  public static void addToInvalidXMLUserObject(TreeNodeWithChangeSupport t, String attributeName, String invalidValue) {
+    if(!t.isSetUserObjects() || !t.containsUserObjectKey(JSBML.INVALID_XML)) {
+      // TODO 2020/03: this is somewhat abusive code
```
@DEVetter Why abusive code ? 
In fact, there was already a method to do that, in org.sbml.jsbml.xml.parsers.AbstractReaderWriter.processInvalidAttribute(String, String, String, String, Object).
You should use this method instead of your method in XMLTools. You could call .AbstractReaderWriter.processInvalidAttribute from XMLTools if you think it would be easier to find (but this should be in the user/developer guide in fact).

niko-rodrigue replied:
Forget about that, I have change the way it will work slightly and I will make the changes myself.
